predicted-latency-scorer crashes on /v1/chat/completions requests

<!-- Please use this template while reporting a bug and provide as much info as possible. Not doing so may result in your bug not being addressed in a timely manner. Thanks!

If the matter is security related, please disclose it privately via https://kubernetes.io/security/
-->

**What happened**:

When sending a `/v1/chat/completions` request with the `predicted-latency-scorer` plugin enabled, the EPP crashes with a nil pointer dereference panic:

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x8 pc=0x1e51b54]

goroutine 734 [running]:
sigs.k8s.io/gateway-api-inference-extension/pkg/epp/framework/plugins/scheduling/scorer/predictedlatency.(*PredictedLatency).generatePredictions(...)
prediction.go:66
```
**What you expected to happen**:

The predicted-latency-scorer plugin should handle both `/v1/completions` and `/v1/chat/completions` requests without crashing.

**How to reproduce it (as minimally and precisely as possible)**:

1. Enable the `predicted-latency-scorer` in the EPP configuration
2. Send a chat completions request
3. EPP crashes immediately with nil pointer dereference (/v1/completions requests work correctly)


**Anything else we need to know?**:
Add a helper method to `LLMRequestBody` that extracts prompt text from either `Completions.Prompt` or `ChatCompletions.Messages`, so latency prediction works correctly for both API types.

**Environment**:
Kubernetes version: v1.32.5
Inference extension version: https://github.com/kubernetes-sigs/gateway-api-inference-extension/commit/74b7a04b76526fd8e2e2d5897e978aa96887a44a

--- a/pkg/epp/framework/interface/scheduling/types.go
+++ b/pkg/epp/framework/interface/scheduling/types.go
@@ -71,6 +71,36 @@ type LLMRequestBody struct {
 	Conversations *ConversationsRequest `json:"conversations,omitempty"`
 }
 
+// PromptText returns a plain-text representation of the prompt from whichever
+// API type is populated, analogous to CacheSalt().
+func (r *LLMRequestBody) PromptText() string {
+	switch {
+	case r.Completions != nil:
+		return r.Completions.Prompt
+	case r.ChatCompletions != nil:
+		var sb strings.Builder
+		for _, msg := range r.ChatCompletions.Messages {
+			text := msg.Content.PlainText()
+			if text != "" {
+				sb.WriteString(text)
+				sb.WriteString(" ")
+			}
+		}
+		return sb.String()
+	case r.Responses != nil:
+		if s, ok := r.Responses.Input.(string); ok {
+			return s
+		}
+		b, _ := json.Marshal(r.Responses.Input)
+		return string(b)
+	case r.Conversations != nil:
+		b, _ := json.Marshal(r.Conversations.Items)
+		return string(b)
+	default:
+		return ""
+	}
+}
+
 func (r *LLMRequestBody) CacheSalt() string {
 	if r.Conversations != nil {
 		return r.Conversations.CacheSalt

--- a/pkg/epp/framework/interface/scheduling/types_test.go
+++ b/pkg/epp/framework/interface/scheduling/types_test.go
@@ -1,0 +1,133 @@
+/*
+Copyright 2025 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package scheduling
+
+import (
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+)
+
+func TestLLMRequestBody_PromptText(t *testing.T) {
+	tests := []struct {
+		name     string
+		body     *LLMRequestBody
+		expected string
+	}{
+		{
+			name: "completions request returns prompt directly",
+			body: &LLMRequestBody{
+				Completions: &CompletionsRequest{
+					Prompt: "What is the meaning of life?",
+				},
+			},
+			expected: "What is the meaning of life?",
+		},
+		{
+			name: "chat completions with single raw message",
+			body: &LLMRequestBody{
+				ChatCompletions: &ChatCompletionsRequest{
+					Messages: []Message{
+						{Role: "user", Content: Content{Raw: "Hello, how are you?"}},
+					},
+				},
+			},
+			expected: "Hello, how are you? ",
+		},
+		{
+			name: "chat completions with multiple messages",
+			body: &LLMRequestBody{
+				ChatCompletions: &ChatCompletionsRequest{
+					Messages: []Message{
+						{Role: "system", Content: Content{Raw: "You are a helpful assistant."}},
+						{Role: "user", Content: Content{Raw: "Tell me a joke."}},
+					},
+				},
+			},
+			expected: "You are a helpful assistant. Tell me a joke. ",
+		},
+		{
+			name: "chat completions with structured content blocks",
+			body: &LLMRequestBody{
+				ChatCompletions: &ChatCompletionsRequest{
+					Messages: []Message{
+						{
+							Role: "user",
+							Content: Content{
+								Structured: []ContentBlock{
+									{Type: "text", Text: "Describe this image:"},
+									{Type: "image_url", ImageURL: ImageBlock{Url: "http://example.com/img.png"}},
+								},
+							},
+						},
+					},
+				},
+			},
+			expected: "Describe this image:  ",
+		},
+		{
+			name: "responses request with string input",
+			body: &LLMRequestBody{
+				Responses: &ResponsesRequest{
+					Input: "Some response input",
+				},
+			},
+			expected: "Some response input",
+		},
+		{
+			name: "responses request with non-string input",
+			body: &LLMRequestBody{
+				Responses: &ResponsesRequest{
+					Input: map[string]interface{}{"key": "value"},
+				},
+			},
+			expected: `{"key":"value"}`,
+		},
+		{
+			name: "conversations request",
+			body: &LLMRequestBody{
+				Conversations: &ConversationsRequest{
+					Items: []ConversationItem{
+						{Type: "message", Role: "user", Content: "Hello"},
+					},
+				},
+			},
+			expected: `[{"type":"message","role":"user","content":"Hello"}]`,
+		},
+		{
+			name:     "empty body returns empty string",
+			body:     &LLMRequestBody{},
+			expected: "",
+		},
+		{
+			name: "chat completions with no messages",
+			body: &LLMRequestBody{
+				ChatCompletions: &ChatCompletionsRequest{
+					Messages: []Message{},
+				},
+			},
+			expected: "",
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			result := tt.body.PromptText()
+			assert.Equal(t, tt.expected, result)
+		})
+	}
+}

--- a/pkg/epp/framework/plugins/scheduling/scorer/predictedlatency/latencypredictor_helper.go
+++ b/pkg/epp/framework/plugins/scheduling/scorer/predictedlatency/latencypredictor_helper.go
@@ -224,7 +224,7 @@ func recordTTFTTrainingData(
 		endpointRoleLabel,
 		targetEndpointMetadata,
 		m,
-		predictedLatencyCtx.schedulingRequest.Body.Completions.Prompt,
+		predictedLatencyCtx.promptText,
 		predictedLatencyCtx.ttft,
 		0, // TTFT training
 		now,
@@ -299,7 +299,7 @@ func processTokenForLatencyPrediction(
 		endpointRoleLabel,
 		targetEndpointMetadata,
 		m,
-		predictedLatencyCtx.schedulingRequest.Body.Completions.Prompt,
+		predictedLatencyCtx.promptText,
 		0, // TTFT not recorded for TPOT
 		latencyMs,
 		now,
@@ -316,7 +316,7 @@ func processTokenForLatencyPrediction(
 			endpointRoleLabel,
 			targetEndpointMetadata,
 			m,
-			predictedLatencyCtx.schedulingRequest.Body.Completions.Prompt,
+			predictedLatencyCtx.promptText,
 			predictedLatencyCtx.generatedTokenCount,
 			0, // TPOT does not use prefix cache score
 		)

--- a/pkg/epp/framework/plugins/scheduling/scorer/predictedlatency/latencypredictor_helper_test.go
+++ b/pkg/epp/framework/plugins/scheduling/scorer/predictedlatency/latencypredictor_helper_test.go
@@ -139,6 +139,36 @@ func TestBulkPredictWithMetrics_WithPredictedLatencyCtx(t *testing.T) {
 	assert.Equal(t, 0.03, results[0].TPOT)
 }
 
+func TestBulkPredictWithMetrics_ChatCompletionsPrompt(t *testing.T) {
+	mp := &mockPredictor{
+		predictions: map[string]*latencypredictor.PredictionResponse{
+			"0.5": {TTFT: 0.5, TPOT: 0.03},
+		},
+	}
+
+	metricsStates := []*fwkdl.Metrics{{KVCacheUsagePercent: 0.5}}
+	pods := []*fwkdl.EndpointMetadata{
+		{NamespacedName: types.NamespacedName{Namespace: "default", Name: "pod1"}},
+	}
+
+	chatBody := &schedulingtypes.LLMRequestBody{
+		ChatCompletions: &schedulingtypes.ChatCompletionsRequest{
+			Messages: []schedulingtypes.Message{
+				{Role: "user", Content: schedulingtypes.Content{Raw: "Hello world"}},
+			},
+		},
+	}
+	prompts := []string{chatBody.PromptText()}
+	generatedTokenCounts := []int{1}
+	prefixCacheScores := []float64{0.0}
+
+	results, err := bulkPredictWithMetrics(context.Background(), nil, mp, metricsStates, "", pods, prompts, generatedTokenCounts, prefixCacheScores)
+
+	assert.NoError(t, err)
+	assert.Len(t, results, 1)
+	assert.Equal(t, 0.5, results[0].TTFT)
+}
+
 func TestBulkPredictWithMetrics_NilMetricsState(t *testing.T) {
 	mockPredictor := &mockPredictor{}
 	metricsStates := []*fwkdl.Metrics{nil} // Nil metrics state

--- a/pkg/epp/framework/plugins/scheduling/scorer/predictedlatency/prediction.go
+++ b/pkg/epp/framework/plugins/scheduling/scorer/predictedlatency/prediction.go
@@ -42,7 +42,7 @@ type endpointPredictionResult struct {
 }
 
 // generatePredictions creates prediction results for all candidate pods
-func (s *PredictedLatency) generatePredictions(ctx context.Context, request *schedulingtypes.LLMRequest, predictedLatencyCtx *predictedLatencyCtx, candidateEndpoints []schedulingtypes.Endpoint) ([]endpointPredictionResult, error) {
+func (s *PredictedLatency) generatePredictions(ctx context.Context, predictedLatencyCtx *predictedLatencyCtx, candidateEndpoints []schedulingtypes.Endpoint) ([]endpointPredictionResult, error) {
 	logger := log.FromContext(ctx)
 	predictions := make([]endpointPredictionResult, 0, len(candidateEndpoints))
 
@@ -63,7 +63,7 @@ func (s *PredictedLatency) generatePredictions(ctx context.Context, request *sch
 
 		metricsStates[i] = endpoint.GetMetrics()
 		targetEndpointsMetadatas[i] = endpoint.GetMetadata()
-		prompts[i] = request.Body.Completions.Prompt
+		prompts[i] = predictedLatencyCtx.promptText
 		generatedTokenCounts[i] = 1
 		prefixCacheScores[i] = prefixCacheScore
 	}

--- a/pkg/epp/framework/plugins/scheduling/scorer/predictedlatency/preparedata_hooks.go
+++ b/pkg/epp/framework/plugins/scheduling/scorer/predictedlatency/preparedata_hooks.go
@@ -51,7 +51,7 @@ func (s *PredictedLatency) PrepareRequestData(ctx context.Context, request *sche
 		}
 		predictedLatencyCtx.prefixCacheScoresForEndpoints[endpoint.GetMetadata().NamespacedName.Name] = prefixCacheScore
 	}
-	predictions, err := s.generatePredictions(ctx, request, predictedLatencyCtx, endpoints)
+	predictions, err := s.generatePredictions(ctx, predictedLatencyCtx, endpoints)
 	if err == nil && len(predictions) == len(endpoints) {
 		s.updateRequestContextWithPredictions(predictedLatencyCtx, predictions)
 		s.updateHasValidPod(ctx, predictedLatencyCtx, endpoints)

--- a/pkg/epp/framework/plugins/scheduling/scorer/predictedlatency/requestcontrol_hooks.go
+++ b/pkg/epp/framework/plugins/scheduling/scorer/predictedlatency/requestcontrol_hooks.go
@@ -67,6 +67,9 @@ type predictedLatencyCtx struct {
 	tpotObservations          []float64
 	predictedTPOTObservations []float64
 
+	// promptText is the cached plain-text prompt, computed once per request.
+	promptText string
+
 	prefixCacheScoresForEndpoints map[string]float64
 
 	// ttftSLO is the target time to first token SLO for the request.
@@ -82,8 +85,13 @@ type predictedLatencyCtx struct {
 }
 
 func newPredictedLatencyContext(request *schedulingtypes.LLMRequest) *predictedLatencyCtx {
+	var prompt string
+	if request.Body != nil {
+		prompt = request.Body.PromptText()
+	}
 	return &predictedLatencyCtx{
 		schedulingRequest:             *request,
+		promptText:                    prompt,
 		lastSeenMetrics:               make(map[string]*fwkdl.Metrics),
 		prefixCacheScoresForEndpoints: make(map[string]float64),
 		predictionsForScheduling:      make(map[string]endpointPredictionResult),

--- a/pkg/epp/framework/plugins/scheduling/scorer/predictedlatency/requestcontrol_hooks_test.go
+++ b/pkg/epp/framework/plugins/scheduling/scorer/predictedlatency/requestcontrol_hooks_test.go
@@ -77,11 +77,31 @@ func TestNewPredictedLatencyContext(t *testing.T) {
 
 	assert.NotNil(t, ctx)
 	assert.Equal(t, *request, ctx.schedulingRequest)
+	assert.Equal(t, "test prompt", ctx.promptText)
 	assert.NotNil(t, ctx.lastSeenMetrics)
 	assert.NotNil(t, ctx.prefixCacheScoresForEndpoints)
 	assert.NotNil(t, ctx.predictionsForScheduling)
 	assert.Empty(t, ctx.lastSeenMetrics)
 	assert.Empty(t, ctx.prefixCacheScoresForEndpoints)
+}
+
+func TestNewPredictedLatencyContext_NilBody(t *testing.T) {
+	request := &schedulingtypes.LLMRequest{
+		Headers: map[string]string{requtil.RequestIdHeaderKey: "test-nil-body"},
+		Body:    nil,
+	}
+	ctx := newPredictedLatencyContext(request)
+
+	assert.NotNil(t, ctx)
+	assert.Empty(t, ctx.promptText)
+}
+
+func TestNewPredictedLatencyContext_ChatCompletionsPrompt(t *testing.T) {
+	request := createTestChatCompletionsLLMRequest("test-chat", 1.0, 0.05)
+	ctx := newPredictedLatencyContext(request)
+
+	assert.NotNil(t, ctx)
+	assert.Equal(t, "You are a helpful assistant. Tell me a joke. ", ctx.promptText)
 }
 
 func TestPredictedLatency_SetAndGetSLOContext(t *testing.T) {
@@ -668,6 +688,7 @@ func TestPredictedLatencyContext_Fields(t *testing.T) {
 	assert.Nil(t, ctx.targetMetadata)
 	assert.Nil(t, ctx.schedulingResult)
 	assert.Nil(t, ctx.tokenSampler)
+	assert.Equal(t, "test prompt", ctx.promptText)
 }
 
 func TestPredictedLatencyContext_UpdateMetrics(t *testing.T) {

--- a/pkg/epp/framework/plugins/scheduling/scorer/predictedlatency/scorer_test.go
+++ b/pkg/epp/framework/plugins/scheduling/scorer/predictedlatency/scorer_test.go
@@ -120,6 +120,25 @@ func createTestEndpoint(name string, kvCacheUsage float64, runningRequestsSize, 
 }
 
 func createTestLLMRequest(reqID string, ttftSLO, tpotSLO float64) *fwksched.LLMRequest {
+	return createTestLLMRequestWithBody(reqID, ttftSLO, tpotSLO, &fwksched.LLMRequestBody{
+		Completions: &fwksched.CompletionsRequest{
+			Prompt: "test prompt",
+		},
+	})
+}
+
+func createTestChatCompletionsLLMRequest(reqID string, ttftSLO, tpotSLO float64) *fwksched.LLMRequest {
+	return createTestLLMRequestWithBody(reqID, ttftSLO, tpotSLO, &fwksched.LLMRequestBody{
+		ChatCompletions: &fwksched.ChatCompletionsRequest{
+			Messages: []fwksched.Message{
+				{Role: "system", Content: fwksched.Content{Raw: "You are a helpful assistant."}},
+				{Role: "user", Content: fwksched.Content{Raw: "Tell me a joke."}},
+			},
+		},
+	})
+}
+
+func createTestLLMRequestWithBody(reqID string, ttftSLO, tpotSLO float64, body *fwksched.LLMRequestBody) *fwksched.LLMRequest {
 	headers := make(map[string]string)
 	headers[requtil.RequestIdHeaderKey] = reqID
 	if ttftSLO > 0 {
@@ -131,11 +150,7 @@ func createTestLLMRequest(reqID string, ttftSLO, tpotSLO float64) *fwksched.LLMR
 
 	return &fwksched.LLMRequest{
 		Headers: headers,
-		Body: &fwksched.LLMRequestBody{
-			Completions: &fwksched.CompletionsRequest{
-				Prompt: "test prompt",
-			},
-		},
+		Body:    body,
 	}
 }
 
@@ -277,6 +292,22 @@ func TestPredictedLatency_Score(t *testing.T) {
 			request:   createTestLLMRequest("test", 1.0, 0.05),
 			endpoints: []fwksched.Endpoint{},
 			// Should return empty scores map
+			expectedScores: map[string]float64{},
+		},
+		{
+			name: "Chat completions request does not panic",
+			predictor: &mockPredictor{
+				predictions: map[string]*latencypredictor.PredictionResponse{
+					"0.5": {TTFT: 0.5, TPOT: 0.03},
+					"0.6": {TTFT: 0.6, TPOT: 0.04},
+				},
+			},
+			strategy: headroomStrategyLeast,
+			request:  createTestChatCompletionsLLMRequest("test-chat", 1.0, 0.05),
+			endpoints: []fwksched.Endpoint{
+				createTestEndpoint("pod1", 0.5, 2, 1),
+				createTestEndpoint("pod2", 0.6, 3, 2),
+			},
 			expectedScores: map[string]float64{},
 		},
 	}

